yt/frontends/nifti/data_structures.py

```diff
@@ -0,0 +1,190 @@
+        self.float_type = np.float64
+        super().__init__(ds, dataset_type)
+
+    def _detect_output_fields(self):
+        self.field_list = [("nifti", "intensity")]
+
+    def _count_grids(self):
+        self.num_grids = 1
+
+    def _parse_index(self):
+        self.grid_left_edge[0][:] = self.ds.domain_left_edge[:]
+        self.grid_right_edge[0][:] = self.ds.domain_right_edge[:]
+        self.grid_dimensions[0][:] = self.ds.domain_dimensions[:]
+        self.grid_particle_count[0][0] = 0
+        self.grid_levels[0][0] = 0
+        self.max_level = 0
+
+    def _populate_grid_objects(self):
+        # only a single grid, no need to loop
+        g = self.grid(0, self, self.grid_levels.flat[0])
+        g._prepare_grid()
+        g._setup_dx()
+        self.grids = np.array([g], dtype="object")
+
+
+class NiftiDataset(Dataset):
+    _index_class = NiftiHierarchy
+    _field_info_class = NiftiFieldInfo
+
+    def __init__(
+        self,
+        filename,
+        dataset_type="nifti",
+        storage_filename=None,
+        units_override=None,
+        unit_system="cgs",
+        default_species_fields=None,
+    ):
+        self.fluid_types += ("nifti",)
+        super().__init__(
+            filename,
+            dataset_type,
+            units_override=units_override,
+            unit_system=unit_system,
+            default_species_fields=default_species_fields,
+        )
+        self.storage_filename = storage_filename
+        # refinement factor between a grid and its subgrid
+        # self.refine_by = 2
+
+    def _set_code_unit_attributes(self):
+        # This is where quantities are created that represent the various
+        # on-disk units.  These are the currently available quantities which
+        # should be set, along with examples of how to set them to standard
+        # values.
+        #
+        self.length_unit = self.quan(1.0, "mm")
+        # self.mass_unit = self.quan(1.0, "g")
+        # self.time_unit = self.quan(1.0, "s")
+        # self.time_unit = self.quan(1.0, "s")
+        #
+        # These can also be set:
+        # self.velocity_unit = self.quan(1.0, "cm/s")
+        # self.magnetic_unit = self.quan(1.0, "gauss")
+        #
+        # If your frontend uses SI EM units, set magnetic units like this
+        # instead:
+        # self.magnetic_unit = self.quan(1.0, "T")
+
+        # this minimalistic implementation fills the requirements for
+        # this frontend to run, change it to make it run _correctly_ !
+        for key, unit in self.__class__.default_units.items():
+            setdefaultattr(self, key, self.quan(1, unit))
+
+    def _parse_parameter_file(self):
+        def is_number(string):
+            try:
+                float(string)
+                return True
+            except ValueError:
+                return False
+
+        img = nib.load(self.filename)
+        parameters = list(str(img.header).split("\n"))
```

Comment:
So I think the *biggest* improvement we could make here is one I didn't realize was possible until I looked at nibabel a bit more.  I think a *lot* of this can be simplified if we use the methods exposed by `img.header` that start with `get_` *or*, if instead of `str(img.header)` we do `dict(img.header)`.  I think casting it to a `dict` should simplify or remove much of this *albeit* at the expense of also turning everything into arrays that maybe should be unpacked.  What we could potentially do is do the dict conversion into a temporary object, and then for any size-1 arrays, unpack them into a scalar value.  (That would also fix some up/down casting issues too.)